### ISSUE TITLE
Error: Headers already sent ?

### ISSUE THREAD

**Issue by proimage** [doublesecretagency/craft-digitaldownload#3]
Craft CMS 3.1.26 & Digital Download 2.1.3

I'm getting this error when passing an Asset stored in an Asset Volume which is a DigitalOcean Spaces instance (I'm not sure yet if it's exclusive to DigitalOcean Spaces assets or not, but the error seems unrelated...):

```
An Error occurred while handling another error:
yii\web\HeadersAlreadySentException: Headers already sent in  on line 0. in C:\Projects\MS\vendor\yiisoft\yii2\web\Response.php:366
Stack trace:
#0 C:\Projects\MS\vendor\yiisoft\yii2\web\Response.php(339): yii\web\Response->sendHeaders()
#1 C:\Projects\MS\vendor\yiisoft\yii2\web\ErrorHandler.php(135): yii\web\Response->send()
#2 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(137): yii\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#3 C:\Projects\MS\vendor\yiisoft\yii2\base\ErrorHandler.php(111): craft\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#4 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(68): yii\base\ErrorHandler->handleException(Object(yii\base\ErrorException))
#5 [internal function]: craft\web\ErrorHandler->handleException(Object(yii\base\ErrorException))
#6 {main}
Previous exception:
yii\base\ErrorException: Cannot modify header information - headers already sent in C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php:238
Stack trace:
#0 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(81): yii\base\ErrorHandler->handleError(2, 'Cannot modify h...', 'E:\\User Data\\Mi...', 238)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'Cannot modify h...', 'E:\\User Data\\Mi...', 238, Array)
#2 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(238): header('Content-Type: a...')
#3 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(196): doublesecretagency\digitaldownload\services\Download->_downloadFile(Object(craft\elements\Asset), 'https://ms...', Array)
#4 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(92): doublesecretagency\digitaldownload\services\Download->_outputFile(Object(doublesecretagency\digitaldownload\models\Link))
#5 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\controllers\DownloadController.php(61): doublesecretagency\digitaldownload\services\Download->startDownload('w7a1yeHVJwmVyL0...')
#6 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\controllers\DownloadController.php(50): doublesecretagency\digitaldownload\controllers\DownloadController->_download('w7a1yeHVJwmVyL0...')
#7 [internal function]: doublesecretagency\digitaldownload\controllers\DownloadController->actionShortPath('w7a1yeHVJwmVyL0...')
#8 C:\Projects\MS\vendor\yiisoft\yii2\base\InlineAction.php(57): call_user_func_array(Array, Array)
#9 C:\Projects\MS\vendor\yiisoft\yii2\base\Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 C:\Projects\MS\vendor\craftcms\cms\src\web\Controller.php(109): yii\base\Controller->runAction('short-path', Array)
#11 C:\Projects\MS\vendor\yiisoft\yii2\base\Module.php(528): craft\web\Controller->runAction('short-path', Array)
#12 C:\Projects\MS\vendor\craftcms\cms\src\web\Application.php(297): yii\base\Module->runAction('digital-downloa...', Array)
#13 C:\Projects\MS\vendor\yiisoft\yii2\web\Application.php(103): craft\web\Application->runAction('digital-downloa...', Array)
#14 C:\Projects\MS\vendor\craftcms\cms\src\web\Application.php(286): yii\web\Application->handleRequest(Object(craft\web\Request))
#15 C:\Projects\MS\vendor\yiisoft\yii2\base\Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#16 C:\Projects\MS\web\index.php(21): yii\base\Application->run()
#17 {main}
```

**Comment by lindseydiloreto** [comment]
Very interesting. Would you mind copying that same file to a local folder, and try pointing Digital Download at the _local_ copy?

I haven't dealt with DigitalOcean Spaces much yet, no idea if that's where the problem is coming from.

**Comment by proimage** [comment]
Done, but I'm still getting effectively the same thing:

```
An Error occurred while handling another error:
yii\web\HeadersAlreadySentException: Headers already sent in  on line 0. in C:\Projects\MS\vendor\yiisoft\yii2\web\Response.php:366
Stack trace:
#0 C:\Projects\MS\vendor\yiisoft\yii2\web\Response.php(339): yii\web\Response->sendHeaders()
#1 C:\Projects\MS\vendor\yiisoft\yii2\web\ErrorHandler.php(135): yii\web\Response->send()
#2 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(137): yii\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#3 C:\Projects\MS\vendor\yiisoft\yii2\base\ErrorHandler.php(111): craft\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#4 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(68): yii\base\ErrorHandler->handleException(Object(yii\base\ErrorException))
#5 [internal function]: craft\web\ErrorHandler->handleException(Object(yii\base\ErrorException))
#6 {main}
Previous exception:
yii\base\ErrorException: Cannot modify header information - headers already sent in C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php:238
Stack trace:
#0 C:\Projects\MS\vendor\craftcms\cms\src\web\ErrorHandler.php(81): yii\base\ErrorHandler->handleError(2, 'Cannot modify h...', 'C:/Projects/MS...', 238)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'Cannot modify h...', 'C:/Projects/MS...', 238, Array)
#2 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(238): header('Content-Type: a...')
#3 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(196): doublesecretagency\digitaldownload\services\Download->_downloadFile(Object(craft\elements\Asset), 'C:/Projects/MS/...', Array)
#4 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php(92): doublesecretagency\digitaldownload\services\Download->_outputFile(Object(doublesecretagency\digitaldownload\models\Link))
#5 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\controllers\DownloadController.php(61): doublesecretagency\digitaldownload\services\Download->startDownload('TQCoGykOIO7AEAN...')
#6 C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\controllers\DownloadController.php(50): doublesecretagency\digitaldownload\controllers\DownloadController->_download('TQCoGykOIO7AEAN...')
#7 [internal function]: doublesecretagency\digitaldownload\controllers\DownloadController->actionShortPath('TQCoGykOIO7AEAN...')
#8 C:\Projects\MS\vendor\yiisoft\yii2\base\InlineAction.php(57): call_user_func_array(Array, Array)
#9 C:\Projects\MS\vendor\yiisoft\yii2\base\Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 C:\Projects\MS\vendor\craftcms\cms\src\web\Controller.php(109): yii\base\Controller->runAction('short-path', Array)
#11 C:\Projects\MS\vendor\yiisoft\yii2\base\Module.php(528): craft\web\Controller->runAction('short-path', Array)
#12 C:\Projects\MS\vendor\craftcms\cms\src\web\Application.php(297): yii\base\Module->runAction('digital-downloa...', Array)
#13 C:\Projects\MS\vendor\yiisoft\yii2\web\Application.php(103): craft\web\Application->runAction('digital-downloa...', Array)
#14 C:\Projects\MS\vendor\craftcms\cms\src\web\Application.php(286): yii\web\Application->handleRequest(Object(craft\web\Request))
#15 C:\Projects\MS\vendor\yiisoft\yii2\base\Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#16 C:\Projects\MS\web\index.php(21): yii\base\Application->run()
#17 {main}
```

**Comment by lindseydiloreto** [comment]
Ok, good to know DigitalOcean Spaces isn't the problem.

Do you get the same issue with other files as well? Or just this one?

Would you mind sending me the file so I can test locally? You can email it to support@doublesecretagency.com.

**Comment by proimage** [comment]
The exact file isn't the problem. Although it's a 27Mb PDF, I've also tried with a < 1Mb PDF as well as a 11kb PNG. None made any difference.

Also, I checked to see if it being on a Windows system was the problem (wouldn't be the first time) by testing on a regular *nix web server, but I'm getting the same error there as well:

```
An Error occurred while handling another error:
yii\web\HeadersAlreadySentException: Headers already sent in  on line 0. in /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/web/Response.php:366
Stack trace:
#0 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/web/Response.php(339): yii\web\Response->sendHeaders()
#1 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/web/ErrorHandler.php(135): yii\web\Response->send()
#2 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/ErrorHandler.php(137): yii\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#3 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/base/ErrorHandler.php(111): craft\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#4 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/ErrorHandler.php(68): yii\base\ErrorHandler->handleException(Object(yii\base\ErrorException))
#5 [internal function]: craft\web\ErrorHandler->handleException(Object(yii\base\ErrorException))
#6 {main}
Previous exception:
yii\base\ErrorException: Cannot modify header information - headers already sent in /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/services/Download.php:238
Stack trace:
#0 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/ErrorHandler.php(81): yii\base\ErrorHandler->handleError(2, 'Cannot modify h...', '/var/www/stage....', 238)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'Cannot modify h...', '/var/www/stage....', 238, Array)
#2 /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/services/Download.php(238): header('Content-Type: a...')
#3 /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/services/Download.php(196): doublesecretagency\digitaldownload\services\Download->_downloadFile(Object(craft\elements\Asset), 'https://dospace...', Array)
#4 /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/services/Download.php(92): doublesecretagency\digitaldownload\services\Download->_outputFile(Object(doublesecretagency\digitaldownload\models\Link))
#5 /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/controllers/DownloadController.php(61): doublesecretagency\digitaldownload\services\Download->startDownload('qmPBxkz1juRpbuD...')
#6 /var/www/stage.example.com/htdocs/vendor/doublesecretagency/craft-digitaldownload/src/controllers/DownloadController.php(50): doublesecretagency\digitaldownload\controllers\DownloadController->_download('qmPBxkz1juRpbuD...')
#7 [internal function]: doublesecretagency\digitaldownload\controllers\DownloadController->actionShortPath('qmPBxkz1juRpbuD...')
#8 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#9 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/Controller.php(109): yii\base\Controller->runAction('short-path', Array)
#11 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('short-path', Array)
#12 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/Application.php(297): yii\base\Module->runAction('digital-downloa...', Array)
#13 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/web/Application.php(103): craft\web\Application->runAction('digital-downloa...', Array)
#14 /var/www/stage.example.com/htdocs/vendor/craftcms/cms/src/web/Application.php(286): yii\web\Application->handleRequest(Object(craft\web\Request))
#15 /var/www/stage.example.com/htdocs/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#16 /var/www/stage.example.com/htdocs/web/index.php(21): yii\base\Application->run()
#17 {main}
```

**Comment by proimage** [comment]
I'll email you the two error logs in `/storage/logs` (unedited). They're as minimal as I can get them—loaded up the page with the download link, deleted all the logs, then clicked "Download".

**Comment by lindseydiloreto** [comment]
Great, thanks!

**Comment by lindseydiloreto** [comment]
Try this... hack the plugin and edit the following lines:

https://github.com/doublesecretagency/craft-digitaldownload/blob/v2/src/services/Download.php#L225-L226

Change this:

```php
// Start with a clean slate
flush();
```

To this:

```php
// Start with a clean slate
ob_flush();
flush();
```

Let me know if that changes anything.

**Comment by lindseydiloreto** [comment]
If that doesn't work, try this...

```php
// Start with a clean slate
ob_clean();
flush();
```

**Comment by proimage** [comment]
Hmm... neither option made a whit of difference. :-/

**Comment by lindseydiloreto** [comment]
Bummer. I'll reply to you via email. 👍

**Comment by cole007** [comment]
Hi 

Same issue here (testing locally) and getting following error:

```
An Error occurred while handling another error:
yii\web\HeadersAlreadySentException: Headers already sent in  on line 0. in /.../deploy/vendor/yiisoft/yii2/web/Response.php:366
```

If I try `ob_clean()` I get a Yii error: `ob_clean(): failed to delete buffer. No buffer to delete` and likewise `ob_flush()` returns: `ob_flush(): failed to flush buffer. No buffer to flush`

**Comment by cole007** [comment]
To clarify context:

- Craft CMS 3.3.2
- PHP 7.1.30

**Comment by cole007** [comment]
Bit more of an update, after tinkering commenting out the `flush()` command in the `_downloadFile` service method seems to resolve the problem.

**Comment by lindseydiloreto** [comment]
Thanks for the update @cole007!

Which line of code did you remove? `flush()` appears in two places in that method.

**Comment by lindseydiloreto** [comment]
@cole007 @proimage Try replacing the entire `_downloadFile` method with the following...

```php
    private function _downloadFile(Asset $asset, string $filePath, array $optionalHeaders = [])
    {
        // Unlimited PHP memory
        App::maxPowerCaptain();

        // https://www.php.net/manual/en/function.flush.php#87807
        // "... turn off pretty much anything that could cause
        //  unwanted output buffering and turn on implicit flush"
        @apache_setenv('no-gzip', 1);
        @ini_set('zlib.output_compression', 0);
        @ini_set('implicit_flush', 1);
        $ob_get_level = ob_get_level();
        for ($i = 0; $i < $ob_get_level; $i++) {
            ob_end_flush();
        }
        ob_implicit_flush(1);

        // Prevent timeouts
        set_time_limit(0);

        // Default headers
        $defaultHeaders = [
            'Content-Type' => 'application/octet-stream',
            'Content-Disposition' => 'attachment; filename="'.$asset->filename.'"',
            'Content-Length' => $asset->size,
        ];

        // Add optional headers
        $headers = array_merge($defaultHeaders, $optionalHeaders);

        // Start with a clean slate
        ob_flush();

        // Open the file
        $file = @fopen($filePath, 'rb');

        // If file doesn't exist, throw an error
        if (!$file) {
            throw new HttpException(500, 'The file you are looking for does not exist.');
        }

        // Set file headers
        foreach ($headers as $name => $value) {
            header("{$name}: {$value}");
        }

        // Size of chunks
        $chunkSize = (1024 * 8);

        // Read the file out one chunk at a time
        while (!feof($file)) {
            echo fread($file, $chunkSize);
            ob_flush(); // Prevent overflow
        }

        // Close file
        fclose($file);

        // Finish up
        ob_end_flush();
        exit;
    }
```

Let me know if that does the trick!

**Comment by cole007** [comment]
Thanks @lindseydiloreto - that seems to have fixed it.
For reference the `flush()` had removed was line 225: https://github.com/doublesecretagency/craft-digitaldownload/blob/v2/src/services/Download.php#L230

**Comment by proimage** [comment]
> Let me know if that does the trick!


Well, it changes the error message, at least! :p
```
PHP Notice – yii\base\ErrorException
ob_flush(): failed to flush buffer. No buffer to flush
1. in C:\Projects\MS\vendor\doublesecretagency\craft-digitaldownload\src\services\Download.php at line 238

229            'Content-Type' => 'application/octet-stream',
230            'Content-Disposition' => 'attachment; filename="'.$asset->filename.'"',
231            'Content-Length' => $asset->size,
232        ];
233 
234        // Add optional headers
235        $headers = array_merge($defaultHeaders, $optionalHeaders);
236 
237        // Start with a clean slate
238        ob_flush();
239 
240        // Open the file
241        $file = @fopen($filePath, 'rb');
242 
243        // If file doesn't exist, throw an error
244        if (!$file) {
245            throw new HttpException(500, 'The file you are looking for does not exist.');
246        }
247
```

There's more, a whole Craft (Yii?) PHP Notice page, but that's the first trace.

**Comment by proimage** [comment]
> Thanks @lindseydiloreto - that seems to have fixed it.
> For reference the `flush()` had removed was line 225: https://github.com/doublesecretagency/craft-digitaldownload/blob/v2/src/services/Download.php#L230

This also allows the obfuscated download to proceed for me. I don't know what the consequences are of not having a `flush()` command there though...

**Comment by lindseydiloreto** [comment]
@proimage @cole007 Good news everyone! I believe this is fixed in `v2.1.5`.

Go ahead and grab the latest version. Let me know if you're still running into any issues.
